### PR TITLE
Add NoResultWarning for empty table in sha

### LINKED ISSUE #1836: sha.query raises ValueError
Hi, 

While using Astroquery, I came across the following error. Other coordinates that I have tried seems to work without any problem. 

``` Python
from astroquery import sha
sha.query(ra = 219.57741, dec = 64.171525, size = 0.001)
```

> ---------------------------------------------------------------------------
> ValueError                                Traceback (most recent call last)
> <ipython-input-57-a82c30798667> in <module>
> ----> 1 sha.query(ra = 219.57741, dec = 64.171525, size = 0.001)
> 
> ~/.local/lib/python3.8/site-packages/astroquery/sha/core.py in query(coord, ra, dec, size, naifid, pid, reqkey, dataset, verbosity, return_response, return_payload)
>     145     # To table
>     146     # transpose data for appropriate table instance handling
> --> 147     t = Table(list(zip(*data)), names=col_names, dtype=dtypes)
>     148     return t
>     149 
> 
> ~/.local/lib/python3.8/site-packages/astropy/table/table.py in __init__(self, data, masked, names, dtype, meta, copy, rows, copy_indices, **kwargs)
>     605         names = [fix_column_name(name) for name in names]
>     606 
> --> 607         self._check_names_dtype(names, dtype, n_cols)
>     608 
>     609         # Finally do the real initialization
> 
> ~/.local/lib/python3.8/site-packages/astropy/table/table.py in _check_names_dtype(self, names, dtype, n_cols)
>     827 
>     828         if len(names) != n_cols or len(dtype) != n_cols:
> --> 829             raise ValueError(
>     830                 'Arguments "names" and "dtype" must match number of columns'
>     831                 .format(inp_str))
> 
> ValueError: Arguments "names" and "dtype" must match number of columns
>

### PR DESCRIPTION
**Pull request by nandita11**
This fixes #1836 
@bsipocz do let me know if any further changes are required. Apologies if something is too wrong. This is my first time contribution to open-source!

### CODE REVIEW

File: astroquery/sha/core.py [hunk 1]
```diff
@@ -145,7 +146,9 @@ def parse_line(line, cs=cs):
     # To table
     # transpose data for appropriate table instance handling
     t = Table(list(zip(*data)), names=col_names, dtype=dtypes)
-    return t
+    if len(data) != len(col_names):
+        warnings.warn(NoResultsWarning("No matching rows were found in the query."))
+        return t
```
Comment (ceb8) [inline astroquery/sha/core.py:151]: The warning is great, but in the no results case we want to return an empty Table, not `t`, and if we aren't in the no results case we do want to return `t`.

**Review by ceb8 — CHANGES_REQUESTED** [review]
Just need to adjust the returns a bit.

**Review by bsipocz — CHANGES_REQUESTED** [review]
This is a good start, but unfortunately doesn't yet provide a fix for #1836 

1) the exception is coming from L147, thus anything added after it has no effect on regression the behaviour

2) when fixing a bug, it's rather useful to test it with the reported bug. E.g. if you run from the bug report, you can test locally whether the fix is working or not:
```
from astroquery import sha 
sha.query(ra = 219.57741, dec = 64.171525, size = 0.001)
```
An advanced version for this approach is to start the bugfix with adding a "regression" test to the test suite, and let CI also run on it. Here in astroquery that is a bit complicated, as we don't run the tests that do real queries for the remote servers, so something that emulates the empty return result will be needed. So let's return to the question of adding a test, once the patch is working.

3) context wise I think for this particular warning it's enough to check `len(data) > 0` and if not raise the warning and return with the empty table, as I can foresee other problematic cases that would result with `len(data) != len(col_names)`, but with some valid return data.

**Comment by nandita11** [comment]
@bsipocz I am having a strange error 
`ImportError: Numpy version 1.16.0 or later must be installed to use Astropy. Numpy is not installed.`
But after installing the branch, it does have `numpy==1.19.2` installed. Because of this, I am not able to test the changes

**Comment by nandita11** [comment]
@bsipocz @ceb8 kindly let me know if any further changes are required. I have tested the given bug

File: astroquery/sha/core.py [hunk 2]
```diff
@@ -144,8 +145,13 @@ def parse_line(line, cs=cs):
...
     # transpose data for appropriate table instance handling
-    t = Table(list(zip(*data)), names=col_names, dtype=dtypes)
-    return t
+    if len(data) > 0:
+        t = Table(list(zip(*data)), names=col_names, dtype=dtypes)
+        return t
+    else:
+        warnings.warn(NoResultsWarning("No matching rows were found in the query."))
+        t = Table()
+        return t
```
Comment (ceb8) [inline astroquery/sha/core.py:154]: Suggest moving the "return t" outside of the if/else.

Comment (nandita11) [reply astroquery/sha/core.py:154]: Sorry. This was indent issue!

**Review by ceb8 — APPROVED** [review]
This looks good. I see that you added a test for this case and then removed it, why did you remove it?

**Comment by nandita11** [comment]
@ceb8 after adding the test, the Travis started to fail. Maybe the way to put the test was wrong on my side

File: astroquery/eso/core.py [hunk 3]
```diff
@@ -701,8 +701,12 @@ def retrieve_data(self, datasets, continuation=False, destination=None,
         if not all(valid_datasets):
             invalid_datasets = [ds for ds, v in zip(datasets_to_download,
                                                     valid_datasets) if not v]
-            raise ValueError("The following data sets were not found on the "
-                             "ESO servers: {0}".format(invalid_datasets))
+            if verbose:
```
Comment (bsipocz) [inline astroquery/eso/core.py:704]: these changes seem to be unrelated to this PR, please remove the commit.

**Review by bsipocz — COMMENTED** [review]
The PR now looks good, pending some minor cleanup of unrelated changes.

**Comment by bsipocz** [comment]
@nandita11 - the test you added is all fine, the only issue with it that it has been added to the mock tests, but we're not mocking an empty table return.

`sha` is unfortunately one of the modules where the test coverage is not great, so adding this as the first remote data test would be a great addition, but given that the other cases are missing, too, it can be part of a follow-up, bigger PR. 

So I would say, we should merge this once the unrelated changes are removed.

File: astroquery/eso/core.py (hunk 3 above)
Comment (nandita11) [reply astroquery/eso/core.py]: Sorry!

**Review by bsipocz — APPROVED** [review]
I'm sorry it took so long to get back to this. I went ahead and rebased, and added a test. Will merge once CI is all green.

**Comment by bsipocz** [comment]
Thanks you @nandita11!
